src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -198,820 +562,1166 @@ private void checkDataLength(int ... lengths) {
-        if (aadBuffer != null) {
+            if (aadBuffer == null) {
+                if (sizeOfAAD == 0 && !aadProcessed) {
+                    aadBuffer = new ByteArrayOutputStream(len);
+                } else {
+                    // update has already been called
+                    throw new IllegalStateException
+                        ("Update has been called; no more AAD data");
+                }
+            }
             aadBuffer.write(src, offset, len);
-        } else {
-            // update has already been called
-            throw new IllegalStateException
-                ("Update has been called; no more AAD data");
         }
-    }
-
-    // Feed the AAD data to GHASH, pad if necessary
-    void processAAD() {
-        if (aadBuffer != null) {
-            if (aadBuffer.size() > 0) {
-                byte[] aad = aadBuffer.toByteArray();
-                sizeOfAAD = aad.length;
 
-                int lastLen = aad.length % AES_BLOCK_SIZE;
-                if (lastLen != 0) {
-                    ghashAllToS.update(aad, 0, aad.length - lastLen);
-                    byte[] padded = expandToOneBlock(aad, aad.length - lastLen,
-                                                     lastLen);
-                    ghashAllToS.update(padded);
-                } else {
-                    ghashAllToS.update(aad);
+        // Feed the AAD data to GHASH, pad if necessary
+        void processAAD() {
+            if (aadBuffer != null) {
+                if (aadBuffer.size() > 0) {
+                    byte[] aad = aadBuffer.toByteArray();
+                    sizeOfAAD = aad.length;
+
+                    int lastLen = aad.length % blockSize;
+                    if (lastLen != 0) {
+                        ghashAllToS.update(aad, 0, aad.length - lastLen);
+                        byte[] padded = expandToOneBlock(aad,
+                            aad.length - lastLen, lastLen, blockSize);
+                        ghashAllToS.update(padded);
+                    } else {
+                        ghashAllToS.update(aad);
+                    }
                 }
+                aadBuffer = null;
             }
-            aadBuffer = null;
+            aadProcessed = true;
         }
-    }
 
-    // Utility to process the last block; used by encryptFinal and decryptFinal
-    void doLastBlock(byte[] in, int inOfs, int len, byte[] out, int outOfs,
-                     boolean isEncrypt) throws IllegalBlockSizeException {
-        byte[] ct;
-        int ctOfs;
-        int ilen = len;  // internal length
+        /**
+         * Process en/decryption all the way to the last block.  It takes both
+         * For input it takes the ibuffer which is wrapped in 'buffer' and 'src'
+         * from doFinal.
+         */
+        int doLastBlock(GCM op, ByteBuffer buffer, ByteBuffer src, ByteBuffer dst) {
+            int resultLen = 0;
+
+            int bLen = getBufferedLength();
```

Comment:
Well, it seems strange to calculate bLen based on 'ibuffer' instead of the specified 'buffer'. For code clarity, it seems better to just use 'buffer'?